### ISSUE TITLE
Alternative serialization formats?

### ISSUE THREAD

**Issue by DeepAnchor** [centrifugal/centrifugo#109]
Hey, this project looks really cool, love what you guys are doing with it!

I was digging through the source and noticed that protobufs were being use to serialize messages in Redis -have you guys considered alternative/pluggable serialization formats for client-server communication as well? 

It would be nice to be able to swap JSON out for flatbuffers, protobuf, etc. since JSON can be quite wasteful bandwidth-wise when you don't need something that's human readable.

**Comment by FZambia** [comment]
Hello @DeepAnchor ! This is quite a big work because of we need to update all our client libraries to support this new serialization format. Do you have a real-life use case where you need this?

**Comment by DeepAnchor** [comment]
I'm evaluating Centrifugo for a realtime multiplayer mobile game that will communicate game state with the back end using a binary protocol. 

I could probably just roll my own using a less feature-full websocket library, but Centrifugo provides some quite nice features out of the box like presence information, channel authorization, etc., built right in.

Having a pluggable serialization format would allow Centrifigo to be a much more generalized websocket server and open up a lot of applications, I think.  But I can understand that it would require a lot of work. Perhaps it can be put on a future wishlist!

**Comment by banks** [comment]
@deepanchor

Just as an FYI as we seem to not communicate this well: centrifugo is not really designed to be a "general purpose websocket server" - websockets are actually an implementation detail is personally prefer we didn't support as there are alternative ways to do server to client push that are as good and with fewer downsides. 

Centrifugo really has no good support for pushing data from client to server - it can be done and will be rebroadcast among other subscribers but for almost any real world app this is not sufficient - there is really no good way to get that data back into your app where it belongs. 

Even for chat apps where you don't care about persistence at all (rare) we had people hit walls with needing analytics info or audit logging and the answer to all of these is: push data from client to your regular app with HTTP and so what you need for persistence etc. and just use centrifugo as a server to web client "push" transport. If rather people saw it as analogous to apple or google push service for web than a general websocket solution. 

If your game really needs two way comma then centrifugo is not a fit for you. 

All that said, this issue is not irrelevant - there might be plausible use-cases for alternative serialisation with only server push semantics, but unless I misunderstood your use-case ids probably not a good fit anyway.

**Comment by FZambia** [comment]
@DeepAnchor yep, we can leave this open for future, not sure though when this will be done – so even if your game requires one-way server-to-client push you probably will finish your game till that moment :)

What I thought recently is gRPC for client-server communication. It's not possible to use GRPC from web browser yet but I saw they do a research in this area. But even without browser support I can't say a lot how gRPC fit Centrifugo – maybe someone with gRPC experience can estimate this?

**Comment by DeepAnchor** [comment]
@banks interesting, thanks for the clarification. I can what you mean now when you say that  Centrifugo fits more into the push service space -the diagram with the one way arrows make a lot more sense now.  
@FZambia you're probably right :) No idea about gRPC stuff, but it seems to be taking over the world right now. Making RPC calls from browsers though... that's something i'd like to see!

**Comment by FZambia** [comment]
@DeepAnchor after refactoring Centrifugo during last 3 months I tried to investigate how difficult it would be to add new serialization format.  And it's generally possible of course. But it's not simple and there is no clear goal in doing such a hard work -  what to do with clients, which format, who will use it. If someone interested in implementing alternative format and can help with this - me and @banks are open to discussion. Until that moment I think we have to close this.

**Comment by FZambia** [comment]
Let's leave it open.
